zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
 	caminfovalid = true;
 }
 
-// Get the average of the values in the cv::Mat depth contained within
-// the supplied bounding rectangle
-double avgOfDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect)
+// Find the median of a continuous cv::Mat
+float findMedianOfMat(cv::Mat mat) {
+	float median = 0;
+	if (mat.isContinuous()) {
+		// copy matrix data to a vector
+		std::vector<float> vec;
+		mat = mat.reshape(0, 1);
+		mat.copyTo(vec);
+		// remove 255 (when this is called, values that are masked are 255)
+		vec.erase(std::remove(vec.begin(), vec.end(), 255), vec.end());
+		// sort vector
+		std::sort(vec.begin(), vec.end()); // forgot this earlier
+		if ((vec.size() % 2) != 0) { // if odd
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median = vec[vec.size()/2];
+		} else { // if even
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			it = vec.begin() + vec.size()/2 - 1;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			median = ((float)median/2.0d);
+		}
+	}
+	return median;
+}
+
+// K_MEANS = k-means, Forgy method of initializing (k-means++ would likely be better, but Forgy is working ok for now)
+// CONTOURS = find depth by finding contours and taking the lowest value of the biggest contour
+// CONTOURS_NON_ADAPTIVE = contours, but using thresholding based on the median instead of adaptive thresholding
+enum DepthCalculationAlgorithm { K_MEANS, CONTOURS, CONTOURS_NON_ADAPTIVE };
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using contour finding
+double contoursDepthMat(const cv::Mat& depth_, const cv::Rect& bound_rect, bool debug = false, bool adaptive = true) {
```

Comment:
Nitpicking - maybe consider swapping the order of debug and adaptive?  I could see more cases where, once code was working, that we'd want to switch between adaptive or not based on e.g. lighting conditions than we would having to turn debug on and off.  Swapping the arg order would let that happen without having to pass in the debug arg in those cases?
Like I said, nitpicking but see what you think.

Reply:
And then I read the rest of the code and saw that debug was always passed in from the calling function...  Never mind :)